pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1094,6 +1086,24 @@ public void updateSchema(Schema schema, boolean reload)
     }
   }
 
+  /**
+   * Helper method to update the schema, or throw SchemaBackwardIncompatibleException when the new schema is not
+   * backward-compatible with the existing schema.
+   */
+  private void updateSchema(Schema schema, Schema oldSchema)
+      throws SchemaBackwardIncompatibleException {
+    schema.updateBooleanFieldsIfNeeded(oldSchema);
+    if (schema.equals(oldSchema)) {
+      LOGGER.info("New schema: {} is the same as the existing schema, not updating it", schema.getSchemaName());
+      return;
+    }
+    if (!schema.isBackwardCompatibleWith(oldSchema)) {
+      throw new SchemaBackwardIncompatibleException(
+          String.format("New schema: %s is not backward-compatible with the existing schema", schema.getSchemaName()));
```

Comment:
is it possible to also return the details of the incompatiblity?

Reply:
We can change the `schema.isBackwardCompatibleWith()` to throw exception instead of returning a boolean. This is out of the scope of this PR, so added to TODO to track that